service/src/main/kotlin/fi/espoo/evaka/pis/controllers/PersonController.kt

```diff
@@ -96,6 +105,76 @@ class PersonController(
+    @PostMapping("/{personId}/duplicate")
+    fun duplicate(
+        db: Database,
+        user: AuthenticatedUser,
+        clock: EvakaClock,
+        @PathVariable personId: PersonId
+    ): PersonId {
+        return db.connect { dbc ->
+                dbc.transaction { tx ->
+                    accessControl.requirePermissionFor(
+                        tx,
+                        user,
+                        clock,
+                        Action.Person.DUPLICATE,
+                        personId
+                    )
+
+                    val duplicateId =
+                        tx.duplicatePerson(personId) ?: throw NotFound("Person $personId not found")
+                    tx.getChild(personId)?.let { child ->
+                        tx.upsertAdditionalInformation(
+                            childId = duplicateId,
+                            data = child.additionalInformation
+                        )
+                    }
+
+                    val parentRelationships =
+                        tx.getChildGuardians(personId).map { guardianId ->
+                            CreateFosterParentRelationshipBody(
+                                childId = duplicateId,
+                                parentId = guardianId,
+                                DateRange(clock.today(), null)
+                            )
+                        } +
+                            tx.getFosterParents(personId).map { relationship ->
+                                CreateFosterParentRelationshipBody(
+                                    childId = duplicateId,
+                                    parentId = relationship.parent.id,
+                                    validDuring = relationship.validDuring
+                                )
+                            }
+                    parentRelationships.forEach { relationship ->
+                        tx.createFosterParentRelationship(relationship)
+                    }
+
+                    val childRelationships =
+                        tx.getGuardianChildIds(personId).map { childId ->
+                            CreateFosterParentRelationshipBody(
+                                childId = childId,
+                                parentId = duplicateId,
+                                validDuring = DateRange(clock.today(), null)
+                            )
+                        } +
+                            tx.getFosterChildren(personId).map { relationship ->
+                                CreateFosterParentRelationshipBody(
+                                    childId = relationship.child.id,
+                                    parentId = duplicateId,
+                                    validDuring = relationship.validDuring
+                                )
+                            }
+                    childRelationships.forEach { relationship ->
+                        tx.createFosterParentRelationship(relationship)
+                    }
```

Comment:
If I understand this correctly, it seems that this method would be 30 % smaller if it didn't support duplicating adults who can be someone's guardian.

Reply:
Fixed